### PR TITLE
Unit-3 Chapter-3 Vision Transformers Transfer learning for classification. 

### PR DESCRIPTION
**Pull request by Anindyadeep**
This PR is contributed by @shreydan @Anindyadeep @alanahmet @hanouticelina @asusevski @sezan92

In this PR, the following things are added:

1. The initial code for classification using transfer learning.
	1. Model: `google/vit-base-patch16-224`: this is the most downloaded and well-known base transformer model.
	2. Dataset: `oxford-iiit pets dataset`: [hf-link](https://huggingface.co/datasets/pcuenq/oxford-pets) - quite a popular dataset for image classification on the HF hub. 
	3. Exclusively used HuggingFace API, that includes `datasets`, `evaluate` and `transformers`
	
2. This initial PR is merged in the forked repo after review by @Anindyadeep @sezan92 and  @alanahmet
3. This also adds the code 

Some questions to the reviewers in the main repo.
1. Should we only prioritize HuggingFace APIs or should we also include custom loops using PyTorch in the end. 
2. For Transfer Learning / Fine-tuning, should we also add segmentation and object detection along with classification? (more explained in this issue)
4. Right now this PR only adds some codes for running, transfer learning for classification. But, how should we structure the folder? Should we follow this examples?
```
1. Transfer Learning
		1. Classification.mdx
		2. Object detection.mdx
		3. segmentation.mdx
	2. Fine-tuning
		1. Classification.mdx
		2. Object detection.mdx
		3. segmentation.mdx
```

Or, should we have follow a structure where we have two files for Transfer learning and fine-tuning and inside that, we include all of the tasks at once. 
```
.
└── course/
    └── chapters/
        ├── chapter3/
        │   ├── Transfer Learning and Fine-tuning/
                        └── transfer-learning.mdx
                        └── fine-tuning.mdx
                        └── knowledge-distillation.mdx
        ├── chapter-n/
        │   └── ...mdx
        └── _toctree.yml
```

5. Right now this PR adds some code for Classification using Transfer Learning, but since this is going to be a tutorial, so how much explaination (in terms of natural words) is needed and diagrams etc. 

Please note: This PR only contains the Jupyter Notebook, based on the feedback, we will start creating `.mdx` files for the same :)

### CODE REVIEW

**Review by merveenoyan — COMMENTED** [review]
Quite end-to-end and nicely put 💜  left some comments on reviewNB to structure the notebook and also added some comments about using transformers 🤗

**Comment by shreydan** [comment]
> Quite end-to-end and nicely put 💜 left some comments on reviewNB to structure the notebook and also added some comments about using transformers 🤗

Thank you very much for the review @merveenoyan! Will keep all your detailed and helpful suggestions in mind, this was an initial PR to understand more about how everything should be arranged.

I will update this notebook accordingly and also add the mdx file. Thank you! 🤗

**Comment by johko** [comment]
Nice Demo @shreydan, as Merve already suggested in some of the comments - feel free to make it a bit more conversational. The focus should be on the code, but there is no harm in writing some more text here and there :slightly_smiling_face: 

Keep it up :+1:

**Comment by shreydan** [comment]
@MKhalusova @johko @merveenoyan I've made further updates to my PR as suggested. 

Please check the image-classification notebooks:
- `notebooks/Unit 3 - Vision Transformers/fine-tuning-multilabel-image-classification.ipynb`
- `notebooks/Unit 3 - Vision Transformers/transfer-learning-image-classification.ipynb`

For transfer-learning:
- updated notebook as per the reviews before
- added pushing model to hub + inference visualization

For fine-tuning:
- new notebook with timm, accelerate, evaluate for multi-label image classification
- detailed comments and explanations on writing training loops
- inference visualization

Based on further reviews, I was thinking to then create a `.mdx` file which contains both transfer learning and fine-tuning parts.

**Comment by merveenoyan** [comment]
@Anindyadeep can you tell me which files are contributed through this PR since I remember I reviewed some of the notebooks in other PRs that seems like a change in this PR. It would be easier for us to review.

**Comment by Anindyadeep** [comment]
> @Anindyadeep can you tell me which files are contributed through this PR since I remember I reviewed some of the notebooks in other PRs that seems like a change in this PR. It would be easier for us to review.

Yeah, so all the sub-topics this includes:

1. Fine-tuning for object detection
2. Lora for Image classification
3. Transfer Learning for Segmentation
4. Knowledge distillation
5. Transfer learning for segmentation
6. Fine-tuning multi class classification 

All of these topics are included in the PR as a whole.

**Comment by shreydan** [comment]
@MKhalusova @merveenoyan @johko 

Sorry for this confusion, this PR is for the **image classification** part under unit-3 vision transformers : fine-tuning and transfer-learning.

We have other PRs for segmentation, object detection, LoRA and Knowledge distillation.

For this PR:

Please check the image-classification notebooks:

    - notebooks/Unit 3 - Vision Transformers/fine-tuning-multilabel-image-classification.ipynb
    - notebooks/Unit 3 - Vision Transformers/transfer-learning-image-classification.ipynb

For image classification, transfer-learning:

 -  updated notebook as per the reviews before
 - added pushing model to hub + inference visualization

For image classification, fine-tuning:

  - new notebook with timm, accelerate, evaluate for multi-label image classification
  - detailed comments and explanations on writing training loops
  - inference visualization

Based on further reviews, I was thinking to then create a .mdx file which contains both transfer learning and fine-tuning parts.

**Comment by merveenoyan** [comment]
@shreydan I think we will not merge this PR until the other PRs are merged to make sure all the content aligns. Moreover, I think @alanahmet's PR already included the first notebook you drafted (transfer learning one) and LoRA one so they're already merged. 
Let us merge the other ones and come back to this if it's ok :)

**Comment by merveenoyan** [comment]
Hey @shreydan 👋 We currently have transfer learning one and LoRA one already merged. Have you renamed the transfer learning one, it should've been in this PR somewhere. 
Moreover, can you check if your PR aligns with the merged notebooks? 
I will check which other PRs you have and merge them and come back to this if it's ok.

**Comment by shreydan** [comment]
@merveenoyan sorry didn't know the transfer learning file was merged, actually I have renamed it to `transfer-learning-image-classification.ipynb` since there were object detection and segmentation tasks as well.

**Comment by merveenoyan** [comment]
@shreydan if you can change the name of the file we can change it in another PR

**Comment by merveenoyan** [comment]
@shreydan we can first wait for other PRs to be merged and then merge this if it's ok for you

**Comment by shreydan** [comment]
> @shreydan if you can change the name of the file we can change it in another PR
@merveenoyan what should I rename the file to? 

@shreydan we can first wait for other PRs to be merged and then merge this if it's ok for you

okay no issues in that. Sorry for this confusion.

**Comment by merveenoyan** [comment]
@Anindyadeep @shreydan I think we can merge this one. If you could sync the changes with this repository and solve merge conflicts I'll give a one last review 🤓

**Comment by Anindyadeep** [comment]
Sure we can do that, and then let's get it merged :)

**Comment by merveenoyan** [comment]
@Anindyadeep do you need help with this?

**Comment by Anindyadeep** [comment]
Hey @merveenoyan, so sorry for the delay, I am bit worked up with some priorities, give me few days I along with @shreydan will solve this very soon :)

**Comment by shreydan** [comment]
hello @merveenoyan sorry for the conflicts, our unit had multiple PRs hence the problems. 1 old notebook was removed and has been updated with 2 new notebooks with regards to ViT image classification (transfer learning & fine tuning). Would really appreciate reviews on these 2 notebooks. Thank you for your patience

**Review by merveenoyan — APPROVED** [review]
I think this PR looks ready to merge 😊 thanks a lot for your patience!

**Comment by merveenoyan** [comment]
@MKhalusova can you skim through the notebooks?

**Review by MKhalusova — APPROVED** [review]
Amazing work! I have no comments to add :)

**Comment by johko** [comment]
Merging :rocket:

**Comment by shreydan** [comment]
@johko @MKhalusova @merveenoyan thank you very much!
